### PR TITLE
Addressing inconsistencies in handling of multiple slashes

### PR DESCRIPTION
**Pull request by JunTaoLuo**
https://github.com/aspnet/StaticFiles/issues/147

Now we will trim all leading slashes (if more than one) and trim all slashes recognized by the underlying file system. This will be consistent with how the slashes elsewhere in the path is treated. I will look into adding some more test variations to increase coverage.

cc @Tratcher

### CODE REVIEW

File: src/Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs [hunk 1]
```diff
@@ -247,24 +248,16 @@ public IDirectoryContents GetDirectoryContents(string subpath)
...
-            // Relative paths starting with a leading slash okay
-            if (filter.StartsWith("/", StringComparison.Ordinal))
-            {
-                filter = filter.Substring(1);
-            }
-
-            // Absolute paths not permitted.
-            if (Path.IsPathRooted(filter))
+            // Absolute paths and paths traversing above root not permitted.
+            if (Path.IsPathRooted(filter) || PathNavigatesAboveRoot(filter))
```
Comment (JunTaoLuo) [inline src/Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs]: Moved the check for navigation above root to after trimming leading slashes which is how we validate for files and directories. Not sure why this was done before trimming previously.

**Comment by Tratcher** [comment]
you meant @blowdart

**Comment by JunTaoLuo** [comment]
cc @BrennanConroy @davidfowl

File: src/Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs [hunk 2]
```diff
@@ -18,6 +18,8 @@ public class PhysicalFileProvider : IFileProvider, IDisposable
         private const string PollingEnvironmentKey = "DOTNET_USE_POLLING_FILE_WATCHER";
         private static readonly char[] _invalidFileNameChars = Path.GetInvalidFileNameChars()
             .Where(c => c != Path.DirectorySeparatorChar && c != Path.AltDirectorySeparatorChar).ToArray();
+        private static readonly char[] _invalidFilterChars = _invalidFileNameChars
```
Comment (BrennanConroy) [inline src/Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs]: Is this right? What about things like `?` and `&` etc.

Comment (JunTaoLuo) [inline src/Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs]: Fixed? It's now:
```
    [0]: 34 '"'
    [1]: 60 '<'
    [2]: 62 '>'
    [3]: 0 '\0'
    [4]: 1 '\u0001'
    [5]: 2 '\u0002'
    [6]: 3 '\u0003'
    [7]: 4 '\u0004'
    [8]: 5 '\u0005'
    [9]: 6 '\u0006'
    [10]: 7 '\a'
    [11]: 8 '\b'
    [12]: 9 '\t'
    [13]: 10 '\n'
    [14]: 11 '\v'
    [15]: 12 '\f'
    [16]: 13 '\r'
    [17]: 14 '\u000e'
    [18]: 15 '\u000f'
    [19]: 16 '\u0010'
    [20]: 17 '\u0011'
    [21]: 18 '\u0012'
    [22]: 19 '\u0013'
    [23]: 20 '\u0014'
    [24]: 21 '\u0015'
    [25]: 22 '\u0016'
    [26]: 23 '\u0017'
    [27]: 24 '\u0018'
    [28]: 25 '\u0019'
    [29]: 26 '\u001a'
    [30]: 27 '\u001b'
    [31]: 28 '\u001c'
    [32]: 29 '\u001d'
    [33]: 30 '\u001e'
    [34]: 31 '\u001f'
    [35]: 58 ':'
```

Comment (Tratcher) [inline src/Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs]: why are we doing this? it does not look related to the leading slashes bug your fixing.

Comment (JunTaoLuo) [inline src/Microsoft.Extensions.FileProviders.Physical/PhysicalFileProvider.cs]: Found an issue when I was adding tests that invalid characters were not being checked for the provider watcher token. Since handling of leading slashes affect those filters, I added the check to make the behaviour consistent across files, directories, and watcher.

**Comment by JunTaoLuo** [comment]
cc @blowdart

**Comment by blowdart** [comment]
:shipit:
